pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh

```diff
@@ -42,7 +42,11 @@ dotnetConfigureHook() {
         if [[ -n "${keepNugetConfig-}" ]]; then
             # If we're keeping the existing configs, we'll add _nix everywhere,
             # in case sources are cleared.
-            dotnet nuget add source "$nugetSource" -n _nix --configfile "$config"
+            if grep -q "$nugetSource" "$config"; then
```

Comment:
Why would this already be there?  It's meant to be a unique temporary path specific to the build.

Reply:
It's technically not strictly necessary for building the final package, but the `fetch-deps` script breaks without it:

```
jasper@VIRTUE:~/repos/nixpkgs$ nix-build -A clps2c-compiler.passthru.fetch-deps
these 4 derivations will be built:
  /nix/store/w6ys796l78fvwiz07afrwjvswiyn2as0-nuget-to-nix.sh.drv
# ...
/nix/store/yxiwf39m7m1v8g51q1g769hy6yyqb06x-CLPS2C-Compiler-1.0.1-fetch-deps
jasper@VIRTUE:~/repos/nixpkgs$ ./result
Running phase: createNugetDirs
The template "NuGet Config" was created successfully.

Processing post-creation actions...


Package source with Name: _nix added successfully.
Running phase: unpackPhase
unpacking source archive /nix/store/vpzlad5blm0vw6268kky8qq0i3ayz3ak-CLPS2C-Compiler
unpacking source archive /nix/store/rnckcc5r1i6i46hlb40s5f0fyf65lf7y-keystone
source root is .
Running phase: patchPhase
applying patch /nix/store/km2a5r65sagbg0qmyq2j19qs0lvmghqk-dont_trim_leading_newline.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/Program.cs
applying patch /nix/store/gbc6yn5gpcgr407lz6pqj0xws6rxp7f7-build_fixes.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/Program.cs
Hunk #1 succeeded at 1162 (offset -4 lines).
Hunk #2 succeeded at 1197 (offset -4 lines).
Hunk #3 succeeded at 1585 (offset -4 lines).
Hunk #4 succeeded at 1803 (offset -4 lines).
Hunk #5 succeeded at 1811 (offset -4 lines).
Hunk #6 succeeded at 1830 (offset -4 lines).
Hunk #7 succeeded at 1908 (offset -4 lines).
Hunk #8 succeeded at 1936 (offset -4 lines).
Hunk #9 succeeded at 1976 (offset -4 lines).
Hunk #10 succeeded at 2004 (offset -4 lines).
Hunk #11 succeeded at 2036 (offset -4 lines).
Hunk #12 succeeded at 2055 (offset -4 lines).
Hunk #13 succeeded at 2068 (offset -4 lines).
Hunk #14 succeeded at 2085 (offset -4 lines).
Hunk #15 succeeded at 2108 (offset -4 lines).
patching file CLPS2C-Compiler/CLPS2C-Compiler/Util.cs
applying patch /nix/store/wba1za7lq7p6b9bhkh4jdm25xpvx6xm3-remove_platformtarget.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj
applying patch /nix/store/2i8da72kkmiyhzb32rh50aj163cf1n32-use_compiled_keystone.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 27 with fuzz 1 (offset -1 lines).
applying patch /nix/store/k2pv26aw09p7afvrmac8pmsim9s18zrv-set_langversion.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 5 with fuzz 2.
applying patch /nix/store/1zvki84w05g5n7kvnidkcy4jn3jwsn34-set_runtimeidentifiers.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 3 with fuzz 1.
applying patch /nix/store/dwfm7i3vin3izb7bqb5xrri3vzdxhrkv-keystone_set_targetframework.patch
patching file keystone/bindings/csharp/Keystone.Net/Keystone.Net.csproj
Running phase: configureNuget
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./keystone/llvm/cmake/config.guess
Running phase: configurePhase
Executing dotnetConfigureHook
error: The name specified has already been added to the list of available package sources. Provide a unique name.
```

Reply:
I think this might have something to do with using `sourceRoot = ".";`.  Could you try it with the source root in a directory?

That's a limitation of the nuget configuration stuff that I'm actually trying to fix right now.

Reply:
Can't seem to get any other value to work, setting it to an arbitrary value gives:
```
$ nix-build -A clps2c-compiler.passthru.fetch-deps && ./result
this derivation will be built:
  /nix/store/k9s43p1kxyakbvwin2f8qj8dmkzfzw98-CLPS2C-Compiler-1.0.1-fetch-deps.drv
building '/nix/store/k9s43p1kxyakbvwin2f8qj8dmkzfzw98-CLPS2C-Compiler-1.0.1-fetch-deps.drv'...
/nix/store/s05akyd1bk2j0i88swhcjy3j9mj16zxm-CLPS2C-Compiler-1.0.1-fetch-deps
Running phase: createNugetDirs
The template "NuGet Config" was created successfully.

Processing post-creation actions...


Package source with Name: _nix added successfully.
Running phase: unpackPhase
unpacking source archive /nix/store/vpzlad5blm0vw6268kky8qq0i3ayz3ak-CLPS2C-Compiler
unpacking source archive /nix/store/rnckcc5r1i6i46hlb40s5f0fyf65lf7y-keystone
source root is asdf
chmod: cannot access 'asdf': No such file or directory
```

And setting it to `CLPS2C-Compiler` will cause patchPhase to fail due to the patch paths no loner lining up.

If I modify the patch file paths (haven't figured out how to get the keystone patch to work since it's now outside of `sourceRoot`, just skipped it for now) will cause this error:
```
$ nix-build -A clps2c-compiler.passthru.fetch-deps && ./result
this derivation will be built:
  /nix/store/7b788c0w24h1cag8cwqr81brvb52z6z6-CLPS2C-Compiler-1.0.1-fetch-deps.drv
building '/nix/store/7b788c0w24h1cag8cwqr81brvb52z6z6-CLPS2C-Compiler-1.0.1-fetch-deps.drv'...
/nix/store/x2mq3acws270d1d3448iriqn0kh8a0f8-CLPS2C-Compiler-1.0.1-fetch-deps
Running phase: createNugetDirs
The template "NuGet Config" was created successfully.

Processing post-creation actions...


Package source with Name: _nix added successfully.
Running phase: unpackPhase
unpacking source archive /nix/store/vpzlad5blm0vw6268kky8qq0i3ayz3ak-CLPS2C-Compiler
unpacking source archive /nix/store/rnckcc5r1i6i46hlb40s5f0fyf65lf7y-keystone
source root is CLPS2C-Compiler
Running phase: patchPhase
applying patch /nix/store/lqj1h8kcqs7x1wi0pi3r4r6r2mb5s59i-dont_trim_leading_newline.patch
patching file CLPS2C-Compiler/Program.cs
applying patch /nix/store/j22539r3bvipkhmhvrn2y6xzp7403aaj-build_fixes.patch
patching file CLPS2C-Compiler/Program.cs
Hunk #1 succeeded at 1162 (offset -4 lines).
Hunk #2 succeeded at 1197 (offset -4 lines).
Hunk #3 succeeded at 1585 (offset -4 lines).
Hunk #4 succeeded at 1803 (offset -4 lines).
Hunk #5 succeeded at 1811 (offset -4 lines).
Hunk #6 succeeded at 1830 (offset -4 lines).
Hunk #7 succeeded at 1908 (offset -4 lines).
Hunk #8 succeeded at 1936 (offset -4 lines).
Hunk #9 succeeded at 1976 (offset -4 lines).
Hunk #10 succeeded at 2004 (offset -4 lines).
Hunk #11 succeeded at 2036 (offset -4 lines).
Hunk #12 succeeded at 2055 (offset -4 lines).
Hunk #13 succeeded at 2068 (offset -4 lines).
Hunk #14 succeeded at 2085 (offset -4 lines).
Hunk #15 succeeded at 2108 (offset -4 lines).
patching file CLPS2C-Compiler/Util.cs
applying patch /nix/store/0jvi1cf7i1zcqc2zl875disxwlpd0yxb-remove_platformtarget.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler.csproj
applying patch /nix/store/53w431v9ns38dqs10kip9fgynziflrx7-use_compiled_keystone.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 27 with fuzz 1 (offset -1 lines).
applying patch /nix/store/x65lhqpywd8yhan1ava4sj1mi7vywmgb-set_langversion.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 5 with fuzz 2.
applying patch /nix/store/ddqg3b2nddgf6lrm2pwdjvmpd605n02a-set_runtimeidentifiers.patch
patching file CLPS2C-Compiler/CLPS2C-Compiler.csproj
Hunk #1 succeeded at 3 with fuzz 1.
Running phase: configureNuget
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
Executing dotnetConfigureHook
  Determining projects to restore...
/nix/store/n0wda6awc5i1a1jmmdsxsr6j7ypbdw6g-dotnet-sdk-8.0.401/sdk/8.0.401/NuGet.targets(174,5): error : Access to the path '/tmp/nix-shell.AOHshk/tmp.Htz4W03LTP/keystone/bindings/csharp/Keystone.Net/obj' is denied. [/tmp/nix-shell.AOHshk/tmp.Htz4W03LTP/CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj]
/nix/store/n0wda6awc5i1a1jmmdsxsr6j7ypbdw6g-dotnet-sdk-8.0.401/sdk/8.0.401/NuGet.targets(174,5): error :   Permission denied [/tmp/nix-shell.AOHshk/tmp.Htz4W03LTP/CLPS2C-Compiler/CLPS2C-Compiler/CLPS2C-Compiler.csproj]
```

Reply:
Sorry, I didn't really understand your source layout with the multiple source inputs.  You could probably create a `source` dir, move both of the unpacked `srcs` into it in `postUnpack`, and then leave `sourceRoot` unset.

There are various other ways you could fix it, but the idea is that `sourceRoot` needs to be one or more directories under the `NIX_BUILD_TOP`.  IMO that's usually a good practice anyway, because various things get dumped in `NIX_BUILD_TOP`, aka `TMPDIR.

Reply:
I don't think this works, you can't leave `sourceRoot` unset if `unpackPhase` produces more than one directory. `postUnpack` doesn't even run.

Reply:
This works, but seems hackier than just fixing `dotnetConfigureHook`

```patch
diff --git a/pkgs/by-name/cl/clps2c-compiler/package.nix b/pkgs/by-name/cl/clps2c-compiler/package.nix
index c09087f5eca9..275cf4c2ea7e 100644
--- a/pkgs/by-name/cl/clps2c-compiler/package.nix
+++ b/pkgs/by-name/cl/clps2c-compiler/package.nix
@@ -38,7 +38,19 @@ buildDotnetModule rec {
     keystone-src
   ];

-  sourceRoot = ".";
+  sourceRoot = "sources";
+
+  preUnpack = ''
+    mkdir sources
+  '';
+
+  postUnpack = ''
+    # Required for mv to work
+    chmod 755 -R CLPS2C-Compiler
+    chmod 755 -R keystone
+    mv CLPS2C-Compiler sources
+    mv keystone sources
+  '';

   patches = [
     ./patches/dont_trim_leading_newline.patch
```

Reply:
#336824 includes my changes to fix the `sourceRoot` problem.

Reply:
Rebased against that PR, seems to work fine now without my hacks.

Reply:
@corngood is there anything else blocking the merge of this PR?